source/material_model/visco_plastic.cc

```diff
@@ -0,0 +1,617 @@
+    {
+      // This function calculates viscosities assuming that all the compositional fields
+      // experience the same strain rate (isostrain).
+
+      // Calculate the square root of the second moment invariant for the deviatoric strain rate tensor.
+      // The first time this function is called (first iteration of first time step)
+      // a specified "reference" strain rate is used as the returned value would
+      // otherwise be zero.
+      const double edot_ii = ( (this->get_timestep_number() == 0 && strain_rate.norm() <= std::numeric_limits<double>::min())
+                               ?
+                               ref_strain_rate
+                               :
+                               std::max(std::sqrt(std::fabs(second_invariant(deviator(strain_rate)))),
+                                        min_strain_rate) );
+
+      // Calculate viscosities for each of the individual compositional phases
+      std::vector<double> composition_viscosities(volume_fractions.size());
+      for (unsigned int j=0; j < volume_fractions.size(); ++j)
+        {
+          // Power law creep equation
+          //    viscosity = 0.5 * A^(-1/n) * edot_ii^((1-n)/n) * d^(-m/n) * exp((E + P*V)/(nRT))
+          // A: prefactor, edot_ii: square root of second invariant of deviatoric strain rate tensor,
+          // d: grain size, m: grain size exponent, E: activation energy, P: pressure,
+          // V; activation volume, n: stress exponent, R: gas constant, T: temperature.
+          // Note: values of A, d, m, E, V and n are distinct for diffusion & dislocation creep
+
+          // Diffusion creep: viscosity is grain size dependent (m!=0) and strain-rate independent (n=1)
+          double viscosity_diffusion = 0.5 * std::pow(prefactors_diffusion[j],-1/stress_exponents_diffusion[j]) *
+                                       std::exp((activation_energies_diffusion[j] + pressure*activation_volumes_diffusion[j])/
+                                                (constants::gas_constant*temperature*stress_exponents_diffusion[j])) *
+                                       std::pow(grain_size, -grain_size_exponents_diffusion[j]);
+
+          // For dislocation creep, viscosity is grain size independent (m=0) and strain-rate dependent (n>1)
+          double viscosity_dislocation = 0.5 * std::pow(prefactors_dislocation[j],-1/stress_exponents_dislocation[j]) *
+                                         std::exp((activation_energies_dislocation[j] + pressure*activation_volumes_dislocation[j])/
+                                                  (constants::gas_constant*temperature*stress_exponents_dislocation[j])) *
+                                         std::pow(edot_ii,((1. - stress_exponents_dislocation[j])/stress_exponents_dislocation[j]));
+
+          // Composite viscosity
+          double viscosity_composite = (viscosity_diffusion * viscosity_dislocation)/(viscosity_diffusion + viscosity_dislocation);
```

Comment:
Thanks for bringing this up, how to handle the dislocation/composite rheology is tricky.  My understanding, which certainly could be wrong, is that most codes follow the approach I took and assume the dislocation strain-rate is the same as the background strain-rate (edot_{ii}).  In regions with high-strain rates, this usually translates to dislocation creep having a much lower viscosity than diffusion creep and the composite viscosity is effectively the same as the dislocation creep viscosity.  The weakest mechanism dominates.  

However, I completely agree this is not necessarily a good assumption and the iterations in `diffusion_dislocation` seem like a reasonable way to approximate the relative contributions.  I often only use a dislocation creep rheology for the lithosphere and need to think about composite rheologies in more detail.

I think a good solution would be to add in an option for determining the relative strain-rate between diffusion/dislocation creep (following method in `diffusion_dislocation`) in the case where a composite rheology is selected.  I'll do some testing first to make sure no odd issues  arise when doing this type of iterations in combination with plastic deformation, but I don't anticipate any.  

Does this sound like a good approach?   I'm all for having as many options as possible with this material model.  It can be really informative to see how a small change to the non-linear rheology can affect short- and long-term dynamics.

Thanks again for bringing this up and reviewing the code!